builtin/providers/aws/resource_aws_instance.go

```diff
@@ -514,44 +530,130 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
-			// IAM profiles can take ~10 seconds to propagate in AWS:
-			//  http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#launch-instance-with-role-console
-			if awsErr.Code() == "InvalidParameterValue" && strings.Contains(awsErr.Message(), "Invalid IAM Instance Profile") {
-				log.Printf("[DEBUG] Invalid IAM Instance Profile referenced, retrying...")
-				time.Sleep(2 * time.Second)
-				continue
-			}
+	instanceID := ""
+	onDemand := true
+
+	if sp := d.Get("spot_price").(string); sp != "" {
+		// FIXME: check that instance type is not t2.micro, t2.small, or t2.medium
+
+		onDemand = false
+		// FIXME: validate spot price casts to a sensible decimal
+
+		spotPlacement := &ec2.SpotPlacement{
+			AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+			GroupName:        aws.String(d.Get("placement_group").(string)),
+		}
+
+		launchSpec := &ec2.RequestSpotLaunchSpecification{
+			Placement:           spotPlacement,
+			BlockDeviceMappings: runOpts.BlockDeviceMappings,
+			ImageID:             runOpts.ImageID,
+			InstanceType:        runOpts.InstanceType,
+			UserData:            runOpts.UserData,
+			EBSOptimized:        runOpts.EBSOptimized,
+			IAMInstanceProfile:  runOpts.IAMInstanceProfile,
+			NetworkInterfaces:   runOpts.NetworkInterfaces,
+		}
+
+		if runOpts.SubnetID != nil {
+			launchSpec.SubnetID = runOpts.SubnetID
+		}
+
+		if runOpts.SecurityGroupIDs != nil {
+			launchSpec.SecurityGroupIDs = runOpts.SecurityGroupIDs
+		}
+
+		if runOpts.SecurityGroups != nil {
+			launchSpec.SecurityGroups = runOpts.SecurityGroups
+		}
+
+		if runOpts.KeyName != nil {
+			launchSpec.KeyName = runOpts.KeyName
+		}
+
+		// Build the spot instance struct
+		spotOpts := &ec2.RequestSpotInstancesInput{
+			SpotPrice:           aws.String(sp), // Required
+			InstanceCount:       aws.Long(1),
+			LaunchSpecification: launchSpec,
+		}
+
+		// Make the spot instance request
+		var spotResp *ec2.RequestSpotInstancesOutput
+		spotResp, err := conn.RequestSpotInstances(spotOpts)
+
+		request_id := *spotResp.SpotInstanceRequests[0].SpotInstanceRequestID
+		d.Set("spot_request_id", request_id)
+
+		spotStateConf := &resource.StateChangeConf{
+			// http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html
+			Pending:    []string{"start", "pending-evaluation", "pending-fulfillment"},
+			Target:     "fulfilled",
+			Refresh:    SpotInstanceStateRefreshFunc(conn, request_id),
+			Timeout:    10 * time.Minute,
+			Delay:      10 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+
+		log.Printf("[DEBUG] waiting for spot bid to resolve... this may take several minutes.")
+		spotRequestRaw, err := spotStateConf.WaitForState()
+		spotRequest := spotRequestRaw.(*ec2.SpotInstanceRequest)
+
+		fallback := d.Get("spot_fallback").(bool)
+
+		if err != nil && !fallback {
+			return fmt.Errorf("Error while waiting for spot request (%s) to resolve: %s", request_id, err)
+		} else if fallback {
```

Comment:
Alright, I'll crest it as a separate resource so that we can see what that
would look like, and DRY things up later

On Wednesday, May 27, 2015, Chris Bednarski <notifications@github.com>
wrote:

> In builtin/providers/aws/resource_aws_instance.go
> <https://github.com/hashicorp/terraform/pull/2095#discussion_r31105170>:
>
> > +			Target:     "fulfilled",
> > +			Refresh:    SpotInstanceStateRefreshFunc(conn, request_id),
> > +			Timeout:    10 * time.Minute,
> > +			Delay:      10 * time.Second,
> > +			MinTimeout: 3 * time.Second,
> > +		}
> > +
> > +		log.Printf("[DEBUG] waiting for spot bid to resolve... this may take several minutes.")
> > +		spotRequestRaw, err := spotStateConf.WaitForState()
> > +		spotRequest := spotRequestRaw.(*ec2.SpotInstanceRequest)
> > +
> > +		fallback := d.Get("spot_fallback").(bool)
> > +
> > +		if err != nil && !fallback {
> > +			return fmt.Errorf("Error while waiting for spot request (%s) to resolve: %s", request_id, err)
> > +		} else if fallback {
>
>  Spot instances take longer to provision, and provisioning is less
> certain. The upside though, is that you can typically save a substantial
> amount of money by using them, and they fit a lot of use cases
> (particularly for ephemeral compute resources) very nicely.
>
> Yeah I see a a ton of value in spot, but those characteristics are pretty
> unpredictable (and different from instances). Terraform aims to be
> declarative. If my plan states that I have 5 instances running and they
> disappear some time later, that declarative model breaks.
>
> On the other hand if I think of the *bid* itself as a resource, the
> declarative mental model stays in tact -- similar to ASGs. I think we can
> DRY by refactoring so don't be afraid to copy-paste some code to get the
> mental model right.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/terraform/pull/2095/files#r31105170>.
>